### ISSUE TITLE
Mongock, when configured with Spring Data, does not respect configured transaction manager

### ISSUE THREAD

**Issue by TheSteinn** [mongock/mongock#676]
## Description

In the [V5 documentation on transactions](https://docs.mongock.io/v5/features/transactions/index.html#configuration), it states that `When mongock.transactionEnabled is true, it enforces native transactions, throwing an exception if the driver is not capable of it.`. Regarding Spring boot ecosystems, it additionally mentions that `in the Springboot ecosystems, the transactions manager needs to be injected in the context.`

The [transaction configuration section of the MongoDB Spring Data Driver page](https://docs.mongock.io/v5/driver/mongodb-springdata/index.html#transactions) never explicitly mentions configuring a `MongoTransactionManager` bean to be available in the Spring context, which may be a documentation error. Though it can be inferred from the various worked examples, for example [this one for Spring Data V4](https://github.com/mongock/mongock-examples/tree/master/mongodb/springboot3-springdata4-jdk17).

The issue I've encountered is that the `connectionDriver` method of `SpringDataMongoV4ContextBase.java` has an autowired dependency to `PlatformTransactionManager`. This method parameter is used in conjunction with the Mongock config to determine if the driver should enable transactions:
```
txManagerOpt.filter(tx -> config.getTransactionEnabled().orElse(true)).ifPresent(tx -> driver.enableTransaction());
```

However, this instance of `PlatformTransactionManager` is not necessarily a `MongoTransactionManager`. For example, in one of my projects that also has JDBC configured, Springboot automatically wires a `JdbcTransactionManager` bean. The `connectionDriver` method had the `JdbcTransactionManager` bean autowired and used this to determine if transactions should be enabled, irrespective of the fact that a `JdbcTransactionManager` could not be used for Mongo transactions. This led to some initial confusion, as the project had not made a `MongoTransactionManager` bean available to the Spring context, and we were unsure why Mongock was not falling over.

This is further compounded by the fact that this autowired `PlatformTransactionManager` bean is not used for the Mongo transactions in `SpringDataMongoV4DriverBase.java` - the method `specificInitialization` explicitly creates a new instance of `MongoTransactionManager`.

So it would appear that the current logic in Mongock v5 for Spring Data is that a `PlatformTransactionManager` of any concrete implementation (not necessarily a Mongo implementation) must be available in the Spring context for transactions to be enabled. However this transaction manager will not be used for Mongo transactions, instead deferring to an instance manually instantiated by Mongock.

Summary of the two issues:

1. `connectionDriver` method in `SpringDataMongoV4ContextBase` does not ensure the `PlatformTransactionManager` bean is applicable to MongoDB
2. Mongock is not using the `MongoTransactionManager` available in the Spring context for transactions of ChangeUnits

The second issue could be considered a non-issue if `SpringDataMongoV4ContextBase` does not use autowired `PlatformTransactionManager` beans, thus deferring enabling transactions solely to the Mongock config. Documentation would need to be updated to explicitly state that Mongock uses its own instance of `MongoTransactionManager`, not the Spring context instance.

## PRIORITY

Normal (since there isn't a severe impact to usage)

### Mongock
- Mongock version: 5.4.0
- Modules involved:  mongock-springboot-v3, mongodb-springdata-v4-driver
- How Mongock is used: annotation approach (via @EnableMongock and application.yml)

### Environment
- Framework and libraries versions. Especially those that affect directly to Mongock(Spring, Spring data, MongoDB driver, etc.): Spring Boot 3.2.3, spring-boot-starter-data-mongodb 3.2.3 (Spring Data MongoDB 4.2.3, MongoDB driver 4.11.1), spring-boot-starter-jdbc 3.2.3 (Spring JDBC 6.1.4)
- Infrastructure: Kubernetes, Docker, SO, etc.: N/A (issue occurring locally)

## Steps to Reproduce

1. Create a Spring Boot app - either manually create a `JdbcTransactionManager` bean, or defer to the Spring autoconfiguration in `DataSourceTransactionManagerAutoConfiguration`. Do not create a `MongoTransactionManager` bean
2. Set `mongock.transaction-enabled: true`
3. Configure Spring Boot app against a MongoDB database with transactions enabled (i.e. replica set)
4. Write a simple migration within the `@Execution` method block

## Behaviour

**Expected behavior:** [What you expect to happen]
Either 
- Mongock errors due to no `MongoTransactionManager` bean being available; OR
- Mongock is successful but does not use a `PlatformTransactionManager` bean to determine if the driver enables transactions

**Actual behavior:** [What actually happens]
- Mongock is successful, but uses a non-Mongo transaction manager for the purposes of checking whether the driver should enable transactions

**How often the bug happens:** [What percentage of the time does it reproduce?]
Always

**Comment by osantana85** [comment]
Hello @TheSteinn , thanks for raising the issue. We'll take a look asap and let you know.

**Comment by osantana85** [comment]
Hello @TheSteinn , we have just released Mongock v5.4.3, fixing this issue. Finally we decided to remove the requirement of PlatformTransactionManager to be injected, because we wasn't being used it (it was necessary in earlier versions).
Thanks.

**Comment by osantana85** [comment]
Hello @TheSteinn , we have received an issue related to this change, it seems we have introduced a breaking change accidentally, because now, if you are not setting `transaction-enabled` property, or its value is `true`, Mongock will execute the process with a transactional context.

In the previous version, if you were not injecting a TransactionManager, the property was ignored.

We probably release a new fix version to rollback the changes, and then we will release a new minor version with a new property that will manage this behavior.

I'll let you know.

**Comment by TheSteinn** [comment]
@osantana85 in my view that is as per documentation:

> When mongock.transactionEnabled is true, it enforces native transactions, throwing an exception if the driver is not capable of it.
> 
> When mongock.transactionEnabled is false, it disables the transactions and ignores the transactionability of the driver.
> 
> When mongock.transactionEnabled is null, it totally delegates on the driver the transactionability of the migration.

And on the MongoDB Spring Data driver page:

> In order to use MongoDB transactions, we need to enable transactions in Mongock(this won't be required in next versions as transactions will be enabled by default) via `driver.enableTransaction();` or `mongock.transaction-enabled=true`.

Though if users have relied on not explicitly setting `transaction-enabled` to `false`, and instead relying on not configuring a transaction manager bean, I can see the change causing a regression in behaviour.

I haven't had much experience with working with non-replica set MongoDB instances, nor MongoDB prior to version 4 (when transactions were added). Does the change cause a problem with either of those two scenarios?

**Comment by ghenadiibatalski** [comment]
@osantana85 
Since v5.4.3 we're also facing the issue if `transaction-enabled` is not explicitly disabled. The workaround is indeed to configure `transaction-enabled` to false. But it would be nice, if the transaction capability would be determined by default.

```
SpringDataMongoV4DriverBase  : Error in Mongock's transaction
io.mongock.api.exception.MongockException: Error in method[DemoDataGeneratorChangeSet.changeSet] : Command failed with error 20 (IllegalOperation): 'Transaction numbers are only allowed on a replica set member or mongos' on server localhost:27017. The full response is {"ok": 0.0, "errmsg": "Transaction numbers are only allowed on a replica set member or mongos", "code": 20, "codeName": "IllegalOperation"}
```

**Comment by osantana85** [comment]
@TheSteinn @ghenadiibatalski  We have just released Mongock version 5.4.4 reverting changes from 5.4.3 (we introduced accidentally a breaking change as I previously said).

We'll work to release a new minor version with a new behavior to fix the original issue raised in this ticket and we'll document it properly.

I'll let you know asap.

Thanks.

**Comment by osantana85** [comment]
Fixed en v5.5.0. We have added a new property `transactional` to prevent breaking changes and also marked as deprecated the existing `transactionEnabled` property.

If you use the old one (`transactionEnabled`), Mongock will work the same way as previous releases, and will use the `TransactionManager` to enabled or not the transactions.

If you use the new one (`transactional`), the `TransactionManager` is not needed anymore, so transactions will be enabled or not based on the value of the property and the driver capabilities.

Hope it helps you.
